### DISCUSSION TITLE
What will be the impact of Airflow redeployment on Celery Executor, K8S Executor and PySpark Job?

### DISCUSSION THREAD

**Discussion by xs2tarunkukreja** [apache/airflow#32275] [Q&A]
We are using Airflow and PySpark Operator on K8S. Our airflow DAGs initiate 3 type of jobs/operations -
1. Job getting executed on Celery executor.
2. Job getting executed on K8S executor.
3. Job getting executed on PySpark Job.
We are using external Postgres DB. Also our code is embedded in airflow image.

Suppose our Airflow DAGs is in running state. And now we redeploy the airflow using Helm Chart. 

What will be the impact of running jobs? Jobs running on Celery executor, Jobs running on K8S executor and Job running as PySpark.

Regards,
Tarun

**Comment by potiuk** [comment]
All the tasks running in Celery will be stopped and restarted. K8S ones will keep running (most likely - this really depens on details of your configuraiton of pod template I believe). When you do a migration between MAJOR versions of airflow, it's recommended to pause all the DAGS before and wait also for K8S jobs to complete.

You should test it. Depending on the operator's capability (and this is something you need to check in the operator's implementation) some of the operators are able to connect to running jobs rather than restart the jobs and some of them handle on_kill() operation and should be able to stop the running jobs (for example Spark). You should really see the code of the operators to see what it does. 

Generally all operators should be "idempotent" which means that it shoud be possible to re-run them and they should produce expected result (but different operators can have different approach there - some realize idempotency by re-running the task from scratch and removing previous data, some will generate stable IDs for the jobs they run and will simply connect to running job or see that the job completed and will not re-rerun the job. Airflow is agnostic to how the operators implement it.

Generally spaking if you care about "gentle" upgrade you can put your celery workers in offline mode before the migration (this can be done with flower or manually sending single TERM signal to your workers. Then the celery workers will stop handling new tasks and complete runnning the task they are running - they will shutdown when all tasks are completed. 

We generally recommend to pause everything and make sure that all tasks are completed before migration - this is the approach that will avoid running some externall jobs multiple times. But details on how ths can be done is up to you if you need it.

**Reply by xs2tarunkukreja** [reply]
Thanks for the information.

**Reply by xs2tarunkukreja** [reply]
@potiuk - I have one more doubt -
Suppose we use Git Sync approach. Now one of our job is running on worker node. In between of the job, git sync occur. What will be the behavior?  Will it refer the new code in between or keep running by using older code?

**Reply by potiuk** [reply]
New code of course. 

This is known feature of airflow DAG s that they should change slowly. Git Sync provides you atomicity of syncs I.e. if your DAG uses another common file in the DAGS folder, those two files will always be synced and run together, but there is no guarantee which "snapshot" of the DAG will be executed at any given time.

Simply Versioning of DAGs is not yet implemented. 

The DRAFT AIP for it has not been completed yet and there is currently no active work happenin on it (but anyone could pick it up).  https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-36+DAG+Versioning

You shoudl avoid changing your DAGs dramatically, so that this becomes a problem, Airflow Assumes that the DAG changes are "slow" and do not change the general behaviour of the tasks and dags - you can alsawasy. If you want to heavlily change your DAG, you should create a new DAG with new dag_id instead.
